README.txt
Welcome to the Kodi-script.video.randomtrailers wiki!

Mission: To resurrect the orphaned and non-functioning Kodi random trailers
screensaver and script.

Status:

Complete rewrite. A lot more function, but bigger. Beta. Fully functional,
but limited testing with Kodi 19 using Python 3.8 & 3.9.

See RELEASE_NOTES.txt for information not contained here.

Major accomplishments:

    * Feature complete
    * Multi-threaded to speed up discovery, improve randomness and eliminate
      wait to download information before each play of remote content

Among the features:

    * Gets trailers from the library, TMDb, TFH and local trailers folder
    * Falls back to TMDb for missing trailers from local database
    * Falls back to TMDb for missing local and TFH movie details

    * One addon, with three extension points:
        1- A trailer discovery module that performs the hard work of finding
           and caching the trailers. Runs as a daemon.
        2- A Frontend is the user facing app. It is used for both screensaver
           and manual launch.
        3- A screensaver service. This thin piece of code gets launched by
           Kodi's screensaver function. It then starts up the frontend.

    * Supports a local cache for trailers and information downloaded from
      remote sites
    * Supports Normalization of audio, mostly to help ghastly trailers/clips
      originating from youtube
    * Filter by Genre, Certification, year, rating, popularity, etc. for
      local, TMDb and TFH
    * Can configure various cache attributes: max size, max number of
      files, max % of disk, delete old files.
    * Plugin runs stand alone or as a screensaver
    * Back-end is a separate long-running service, reducing startup time
      and high startup cost
    * User ability to add current trailer to a playlist mostly to flag
      movies that they would like to watch, or note for any reason
    * Internationalized (limited testing). All messages use translation
      system. Country code and language used in remote movie queries
    * Genre information is loaded from .xml files allowing customization
    * Certification/Rating information is loaded from .xml files to allow
      different rules and names by country. (Kodi appears to have some
      limitations in this area: ratings do not include country info.)

    Support for Text to Speech plugin (under development)


changelog.txt
v2.0.7 (July, 04, 2021)
 - Beta 1 candidate
 - Major refactoring of 'dict' movie structure to classes (wrapping dict).
   - Large, but straightforward changes
   - Allowed changes to delay discovery of data until actually needed
   - Reduced startup time, cpu load and memory requirements
   - Simpler, allows for less duplicate code
 - Reduced volume of remote data stored on disk:
   - Reading/Writing large caches very expensive and done too frequently
   - Instead of saving raw data from TMDb, etc. save digested, RandomTrailers version of data
   - Save minimal info required for an index and lookup full data later
   - In cases where lost changes were self healing, reduce how often written to disk
 - Realized that some xbmc.executebuiltin calls incorrectly had XBMC prefix to call
 - Removed change volume feature since Kodi does not provide ability to query volume

v.2.0.4 (February, 04,2021)
 - Alpha  candidate
 - Nearly complete rewrite
   - Randomly plays trailers for local movies, from a local folder and remote trailers from TMDb and TFH.
   - Ability to locally cache remote trailers
   - Ability to normalize the audio of remote and local trailers in order to reduce overly-loud or quiet trailers, commonly found from fan-generated trailers on youtube.
   - Cache downloaded metadata to reduce network traffic
   - Filtering for Date, Genre, Certification
   - Ability to save movie for currently playing trailer to playlists
   - Front-end plugin provides UI to play trailers. Exits when you stop playing trailers
   - Back-end plugin handles the high-cost of prepairing trailers for view and continues to download trailers in the background even when front-end is not active.
   - Supports Text-to-Speech plugin (still under development).
   - XML files define genres and certifications based on country or language.
 
v.1.1.17 (April 05,2014)
 - fixed issue causing crash with tmdb selected, and layout error in tmdb settings

v.1.1.16 (April 02,2014)
 - fixed issue with crashing on startup when library movie has no trailer and we try to find one from tmdb.
 
v.1.1.15 (March 08,2014)
 - fixed issues with rating limits with unrated movies.
 - now pauses trailer when displaying info
 - transparent background for info shows dimmed paused trailer.

v.1.1.14 (March 06,2014)
 - optimized trailer search on tmdb when rating limits applied
 - fixed issue where search count not displaying correctly for tmdb searches
 
v.1.1.13 (March 03,2014)
 - updated filter feature to filter by rating limit for all trailers (perviously only iTunes trailers were filtered)

v.1.1.12 (February 12,2014)
- added feature to play trailers in groups of n with x amount of delay between groups.

v.1.1.11 (February 10,2014)
- added feature to filter tmdb trailers by New DVD Releases

v.1.10 (February, 07, 2014)
- fixed issue playing library trailers

v.1.1.9 (February, 05,2014)
- fixed issue causing crash for not standard ascii characters. Fix EOL

v.1.1.8 (January, 11, 2014:
- added ability to play trailers from You Tube. Trailers are located using themoviedb.org.
- added CouchPotato integration, while watching trailer pressing Q puts current movie in you CouchPotato Queue.

v.1.1.7 (January, 5, 2014) :
- added feature to display trailer info for iTunes trailers.
-cleaned up and optimized code
-moved hide info option from library to general settings

v.1.1.6 (December 31,2013) :
- added feature to select type of trailers form iTunes, Coming Soon, Just Added, Most Popular,  Exclusive and All

v.1.1.5 (Decembeor 30,2013) :
- added feature to filter iTunes trailers. you can filter by selected genre(s), and by rating limit. so only play family movies that are rated G for example.

v.1.1.4 (December 29, 2013) :
 - added ability to play random trailers from multiple sources
 
v.1.1.3 (December 28, 2013) :
 - added ability to play random trailers form apple itunes
 - reworked options screens to reflect multiple trailer sources
 
v.1.1.2 (December 26, 2013) :
 - added feature to adjust volume during trailer playback
 
v.1.1.1 (December 09, 2013) :
 - removed font changes
 
v.1.1.0 (November 27, 2013) :
 - updates for gothum compatibility
 - added ability to play trailers from a folder rather than the users library
 - fixed issue where number of trailers incorrect if hide watched movies selected.
 - trailers will now not repeat until all trailers have been played.
 - fixed issue where trailers would start playing while music is playing
 
v.1.0.12 (November 25, 2013) :
 - added feature if movie title hidden pressing m displays the title for 2 seconds
 
v.1.0.11 (November 25, 2013) :
 - added option to hide movie title when playing
 - removed malfunctioning password options
 
v.1.0.9 (july 26, 2013) :
 - localized strings for dialogs
 
v1.0.8 (July 25, 2013) :
 - added option to force full screen
 
v1.0.7 (June 16, 2013) :
 - merged program, service and screensaver into one addon
 - added option to set number of trailers to 0 for continous playback.

v1.0.6 (June 14, 2013) :
  - added arrow controls for remotes without keyboards
      UP = show info
      DOWN = hide info
      RIGHT = next trailer
      LEFT = exit trailers
      
v1.0.5 (June 12, 2013) :
  - fixed module import issues. fixed font issues
  
v1.0.4 (June 10, 2013) :
  - improved movie information screen
  - added feature and setting to exclude recently watched movies from trailers. Set to number of days after movie
    has been played before including in trailers
  
v1.0.3 (June 4, 2013) :
  - added setting to hide movie info during playback of trailer
    press I during playback of trailer to toggle info display
  
v1.0.2 (June 3, 2013) :
  - added feature to allow playing movie of currently playing trailer by pressing enter while trailer is playing

actions/actionMap.py
'''
Created on Feb 6, 2019

@author: fbacher
'''
from __future__ import print_function
from __future__ import division
from __future__ import absolute_import
from __future__ import unicode_literals
from future import standard_library
standard_library.install_aliases()
from builtins import str
from builtins import range
from builtins import unicode
from xml.dom.minidom import Node
from multiprocessing.pool import ThreadPool
import datetime
import json
import os
import queue
import random
import re
import requests
import resource
import sys
import threading
import time
import traceback
import urllib.request
import urllib.parse
#from kodi_six import xbmc, xbmcaddon, xbmcgui, xbmcplugin, xbmcvfs
import xbmc
import xbmcaddon
import xbmcgui
import xbmcvfs
import xbmcplugin
import xbmcaddon
#import xbmcwsgi
import xbmcdrm
import xml.dom.minidom
import string


class Action:
    # Values came from xbmcgui

    def __init__(self):
        actionMap = {u'ACTION_ANALOG_FORWARD': 113,
                     u'ACTION_ANALOG_MOVE': 49,
                     u'ACTION_ANALOG_MOVE_X_LEFT': 601,
                     u'ACTION_ANALOG_MOVE_X_RIGHT': 602,
                     u'ACTION_ANALOG_MOVE_Y_DOWN': 604,
                     u'ACTION_ANALOG_MOVE_Y_UP': 603,
                     u'ACTION_ANALOG_REWIND': 114,
                     u'ACTION_ANALOG_SEEK_BACK': 125,
                     u'ACTION_ANALOG_SEEK_FORWARD': 124,
                     u'ACTION_ASPECT_RATIO': 19,
                     u'ACTION_AUDIO_DELAY': 161,
                     u'ACTION_AUDIO_DELAY_MIN': 54,
                     u'ACTION_AUDIO_DELAY_PLUS': 55,
                     u'ACTION_AUDIO_NEXT_LANGUAGE': 56,
                     u'ACTION_BACKSPACE': 110,
                     u'ACTION_BIG_STEP_BACK': 23,
                     u'ACTION_BIG_STEP_FORWARD': 22,
                     u'ACTION_BROWSE_SUBTITLE': 247,
                     u'ACTION_BUILT_IN_FUNCTION': 122,
                     u'ACTION_CALIBRATE_RESET': 48,
                     u'ACTION_CALIBRATE_SWAP_ARROWS': 47,
                     u'ACTION_CHANGE_RESOLUTION': 57,
                     u'ACTION_CHANNEL_DOWN': 185,
                     u'ACTION_CHANNEL_NUMBER_SEP': 192,
                     u'ACTION_CHANNEL_SWITCH': 183,
                     u'ACTION_CHANNEL_UP': 184,
                     u'ACTION_CHAPTER_OR_BIG_STEP_BACK': 98,
                     u'ACTION_CHAPTER_OR_BIG_STEP_FORWARD': 97,
                     u'ACTION_CONTEXT_MENU': 117,
                     u'ACTION_COPY_ITEM': 81,
                     u'ACTION_CREATE_BOOKMARK': 96,
                     u'ACTION_CREATE_EPISODE_BOOKMARK': 95,
                     u'ACTION_CURSOR_LEFT': 120,
                     u'ACTION_CURSOR_RIGHT': 121,
                     u'ACTION_CYCLE_SUBTITLE': 99,
                     u'ACTION_DECREASE_PAR': 220,
                     u'ACTION_DECREASE_RATING': 137,
                     u'ACTION_DELETE_ITEM': 80,
                     u'ACTION_ENTER': 135,
                     u'ACTION_ERROR': 998,
                     u'ACTION_FILTER': 233,
                     u'ACTION_FILTER_CLEAR': 150,
                     u'ACTION_FILTER_SMS2': 151,
                     u'ACTION_FILTER_SMS3': 152,
                     u'ACTION_FILTER_SMS4': 153,
                     u'ACTION_FILTER_SMS5': 154,
                     u'ACTION_FILTER_SMS6': 155,
                     u'ACTION_FILTER_SMS7': 156,
                     u'ACTION_FILTER_SMS8': 157,
                     u'ACTION_FILTER_SMS9': 158,
                     u'ACTION_FIRST_PAGE': 159,
                     u'ACTION_FORWARD': 16,
                     u'ACTION_GESTURE_ABORT': 505,
                     u'ACTION_GESTURE_BEGIN': 501,
                     u'ACTION_GESTURE_END': 599,
                     u'ACTION_GESTURE_NOTIFY': 500,
                     u'ACTION_GESTURE_PAN': 504,
                     u'ACTION_GESTURE_ROTATE': 503,
                     u'ACTION_GESTURE_SWIPE_DOWN': 541,
                     u'ACTION_GESTURE_SWIPE_DOWN_TEN': 550,
                     u'ACTION_GESTURE_SWIPE_LEFT': 511,
                     u'ACTION_GESTURE_SWIPE_LEFT_TEN': 520,
                     u'ACTION_GESTURE_SWIPE_RIGHT': 521,
                     u'ACTION_GESTURE_SWIPE_RIGHT_TEN': 530,
                     u'ACTION_GESTURE_SWIPE_UP': 531,
                     u'ACTION_GESTURE_SWIPE_UP_TEN': 540,
                     u'ACTION_GESTURE_ZOOM': 502,
                     u'ACTION_GUIPROFILE_BEGIN': 204,
                     u'ACTION_HIGHLIGHT_ITEM': 8,
                     u'ACTION_INCREASE_PAR': 219,
                     u'ACTION_INCREASE_RATING': 136,
                     u'ACTION_INPUT_TEXT': 244,
                     u'ACTION_JUMP_SMS2': 142,
                     u'ACTION_JUMP_SMS3': 143,
                     u'ACTION_JUMP_SMS4': 144,
                     u'ACTION_JUMP_SMS5': 145,
                     u'ACTION_JUMP_SMS6': 146,
                     u'ACTION_JUMP_SMS7': 147,
                     u'ACTION_JUMP_SMS8': 148,
                     u'ACTION_JUMP_SMS9': 149,
                     u'ACTION_LAST_PAGE': 160,
                     u'ACTION_MENU': 163,
                     u'ACTION_MOUSE_DOUBLE_CLICK': 103,
                     u'ACTION_MOUSE_DRAG': 106,
                     u'ACTION_MOUSE_END': 109,
                     u'ACTION_MOUSE_LEFT_CLICK': 100,
                     u'ACTION_MOUSE_LONG_CLICK': 108,
                     u'ACTION_MOUSE_MIDDLE_CLICK': 102,
                     u'ACTION_MOUSE_MOVE': 107,
                     u'ACTION_MOUSE_RIGHT_CLICK': 101,
                     u'ACTION_MOUSE_START': 100,
                     u'ACTION_MOUSE_WHEEL_DOWN': 105,
                     u'ACTION_MOUSE_WHEEL_UP': 104,
                     u'ACTION_MOVE_DOWN': 4,
                     u'ACTION_MOVE_ITEM': 82,
                     u'ACTION_MOVE_ITEM_DOWN': 116,
                     u'ACTION_MOVE_ITEM_UP': 115,
                     u'ACTION_MOVE_LEFT': 1,
                     u'ACTION_MOVE_RIGHT': 2,
                     u'ACTION_MOVE_UP': 3,
                     u'ACTION_MUTE': 91,
                     u'ACTION_NAV_BACK': 92,
                     u'ACTION_NEXT_CHANNELGROUP': 186,
                     u'ACTION_NEXT_CONTROL': 181,
                     u'ACTION_NEXT_ITEM': 14,
                     u'ACTION_NEXT_LETTER': 140,
                     u'ACTION_NEXT_PICTURE': 28,
                     u'ACTION_NEXT_SCENE': 138,
                     u'ACTION_NEXT_SUBTITLE': 26,
                     u'ACTION_NONE': 0,
                     u'ACTION_NOOP': 999,
                     u'ACTION_PAGE_DOWN': 6,
                     u'ACTION_PAGE_UP': 5,
                     u'ACTION_PARENT_DIR': 9,
                     u'ACTION_PASTE': 180,
                     u'ACTION_PAUSE': 12,
                     u'ACTION_PLAYER_DEBUG': 27,
                     u'ACTION_PLAYER_FORWARD': 77,
                     u'ACTION_PLAYER_PLAY': 79,
                     u'ACTION_PLAYER_PLAYPAUSE': 229,
                     u'ACTION_PLAYER_PROCESS_INFO': 69,
                     u'ACTION_PLAYER_PROGRAM_SELECT': 70,
                     u'ACTION_PLAYER_RESET': 248,
                     u'ACTION_PLAYER_RESOLUTION_SELECT': 71,
                     u'ACTION_PLAYER_REWIND': 78,
                     u'ACTION_PREVIOUS_CHANNELGROUP': 187,
                     u'ACTION_PREVIOUS_MENU': 10,
                     u'ACTION_PREV_CONTROL': 182,
                     u'ACTION_PREV_ITEM': 15,
                     u'ACTION_PREV_LETTER': 141,
                     u'ACTION_PREV_PICTURE': 29,
                     u'ACTION_PREV_SCENE': 139,
                     u'ACTION_PVR_PLAY': 188,
                     u'ACTION_PVR_PLAY_RADIO': 190,
                     u'ACTION_PVR_PLAY_TV': 189,
                     u'ACTION_PVR_SHOW_TIMER_RULE': 191,
                     u'ACTION_QUEUE_ITEM': 34,
                     u'ACTION_QUEUE_ITEM_NEXT': 251,
                     u'ACTION_RECORD': 170,
                     u'ACTION_RELOAD_KEYMAPS': 203,
                     u'ACTION_REMOVE_ITEM': 35,
                     u'ACTION_RENAME_ITEM': 87,
                     u'ACTION_REWIND': 17,
                     u'ACTION_ROTATE_PICTURE_CCW': 51,
                     u'ACTION_ROTATE_PICTURE_CW': 50,
                     u'ACTION_SCAN_ITEM': 201,
                     u'ACTION_SCROLL_DOWN': 112,
                     u'ACTION_SCROLL_UP': 111,
                     u'ACTION_SELECT_ITEM': 7,
                     u'ACTION_SETTINGS_LEVEL_CHANGE': 242,
                     u'ACTION_SETTINGS_RESET': 241,
                     u'ACTION_SET_RATING': 164,
                     u'ACTION_SHIFT': 118,
                     u'ACTION_SHOW_FULLSCREEN': 36,
                     u'ACTION_SHOW_GUI': 18,
                     u'ACTION_SHOW_INFO': 11,
                     u'ACTION_SHOW_OSD': 24,
                     u'ACTION_SHOW_OSD_TIME': 123,
                     u'ACTION_SHOW_PLAYLIST': 33,
                     u'ACTION_SHOW_SUBTITLES': 25,
                     u'ACTION_SHOW_VIDEOMENU': 134,
                     u'ACTION_SMALL_STEP_BACK': 76,
                     u'ACTION_STEP_BACK': 21,
                     u'ACTION_STEP_FORWARD': 20,
                     u'ACTION_STEREOMODE_NEXT': 235,
                     u'ACTION_STEREOMODE_PREVIOUS': 236,
                     u'ACTION_STEREOMODE_SELECT': 238,
                     u'ACTION_STEREOMODE_SET': 240,
                     u'ACTION_STEREOMODE_TOGGLE': 237,
                     u'ACTION_STEREOMODE_TOMONO': 239,
                     u'ACTION_STOP': 13,
                     u'ACTION_SUBTITLE_ALIGN': 232,
                     u'ACTION_SUBTITLE_DELAY': 162,
                     u'ACTION_SUBTITLE_DELAY_MIN': 52,
                     u'ACTION_SUBTITLE_DELAY_PLUS': 53,
                     u'ACTION_SUBTITLE_VSHIFT_DOWN': 231,
                     u'ACTION_SUBTITLE_VSHIFT_UP': 230,
                     u'ACTION_SWITCH_PLAYER': 234,
                     u'ACTION_SYMBOLS': 119,
                     u'ACTION_TAKE_SCREENSHOT': 85,
                     u'ACTION_TELETEXT_BLUE': 218,
                     u'ACTION_TELETEXT_GREEN': 216,
                     u'ACTION_TELETEXT_RED': 215,
                     u'ACTION_TELETEXT_YELLOW': 217,
                     u'ACTION_TOGGLE_COMMSKIP': 246,
                     u'ACTION_TOGGLE_DIGITAL_ANALOG': 202,
                     u'ACTION_TOGGLE_FONT': 249,
                     u'ACTION_TOGGLE_FULLSCREEN': 199,
                     u'ACTION_TOGGLE_SOURCE_DEST': 32,
                     u'ACTION_TOGGLE_WATCHED': 200,
                     u'ACTION_TOUCH_LONGPRESS': 411,
                     u'ACTION_TOUCH_LONGPRESS_TEN': 420,
                     u'ACTION_TOUCH_TAP': 401,
                     u'ACTION_TOUCH_TAP_TEN': 410,
                     u'ACTION_TRIGGER_OSD': 243,
                     u'ACTION_VIDEO_NEXT_STREAM': 250,
                     u'ACTION_VIS_PRESET_LOCK': 130,
                     u'ACTION_VIS_PRESET_NEXT': 128,
                     u'ACTION_VIS_PRESET_PREV': 129,
                     u'ACTION_VIS_PRESET_RANDOM': 131,
                     u'ACTION_VIS_PRESET_SHOW': 126,
                     u'ACTION_VIS_RATE_PRESET_MINUS': 133,
                     u'ACTION_VIS_RATE_PRESET_PLUS': 132,
                     u'ACTION_VOICE_RECOGNIZE': 300,
                     u'ACTION_VOLAMP': 90,
                     u'ACTION_VOLAMP_DOWN': 94,
                     u'ACTION_VOLAMP_UP': 93,
                     u'ACTION_VOLUME_DOWN': 89,
                     u'ACTION_VOLUME_SET': 245,
                     u'ACTION_VOLUME_UP': 88,
                     u'ACTION_VSHIFT_DOWN': 228,
                     u'ACTION_VSHIFT_UP': 227,
                     u'ACTION_ZOOM_IN': 31,
                     u'ACTION_ZOOM_LEVEL_1': 38,
                     u'ACTION_ZOOM_LEVEL_2': 39,
                     u'ACTION_ZOOM_LEVEL_3': 40,
                     u'ACTION_ZOOM_LEVEL_4': 41,
                     u'ACTION_ZOOM_LEVEL_5': 42,
                     u'ACTION_ZOOM_LEVEL_6': 43,
                     u'ACTION_ZOOM_LEVEL_7': 44,
                     u'ACTION_ZOOM_LEVEL_8': 45,
                     u'ACTION_ZOOM_LEVEL_9': 46,
                     u'ACTION_ZOOM_LEVEL_NORMAL': 37,
                     u'ACTION_ZOOM_OUT': 30}

        '''
            u'ALPHANUM_HIDE_INPUT' : 2,
            u'CONTROL_TEXT_OFFSET_X' : 10,
            u'CONTROL_TEXT_OFFSET_Y' : 2,
            u'HORIZONTAL' : 0,
            u'ICON_OVERLAY_HD' : 6,
            u'ICON_OVERLAY_LOCKED' : 3,
            u'ICON_OVERLAY_NONE' : 0,
            u'ICON_OVERLAY_RAR' : 1,
            u'ICON_OVERLAY_UNWATCHED' : 4,
            u'ICON_OVERLAY_WATCHED' : 5,
            u'ICON_OVERLAY_ZIP' : 2,
            u'ICON_TYPE_FILES' : 106,
            u'ICON_TYPE_MUSIC' : 103,
            u'ICON_TYPE_NONE' : 101,
            u'ICON_TYPE_PICTURES' : 104,
            u'ICON_TYPE_PROGRAMS' : 102,
            u'ICON_TYPE_SETTINGS' : 109,
            u'ICON_TYPE_VIDEOS' : 105,
            u'ICON_TYPE_WEATHER' : 107
        '''

        '''
            u'INPUT_ALPHANUM' : 0,
            u'INPUT_DATE' : 2,
            u'INPUT_IPADDRESS' : 4,
            u'INPUT_NUMERIC' : 1,
            u'INPUT_PASSWORD' : 5,
                u'INPUT_TIME' : 3,
                u'INPUT_TYPE_DATE' : 4,
                u'INPUT_TYPE_IPADDRESS' : 5,
                u'INPUT_TYPE_NUMBER' : 1,
                u'INPUT_TYPE_PASSWORD' : 6,
                u'INPUT_TYPE_PASSWORD_MD5' : 7,
                u'INPUT_TYPE_SECONDS' : 2,
                u'INPUT_TYPE_TEXT' : 0,
            u'INPUT_TYPE_TIME' : 3
        '''

        '''
                u'KEY_APPCOMMAND' : 53248,
                u'KEY_ASCII' : 61696,
        '''

        keyButtonMap = {
            u'KEY_BUTTON_A': 256,
            u'KEY_BUTTON_B': 257,
            u'KEY_BUTTON_BACK': 275,
            u'KEY_BUTTON_BLACK': 260,
            u'KEY_BUTTON_DPAD_DOWN': 271,
            u'KEY_BUTTON_DPAD_LEFT': 272,
            u'KEY_BUTTON_DPAD_RIGHT': 273,
            u'KEY_BUTTON_DPAD_UP': 270,
            u'KEY_BUTTON_LEFT_ANALOG_TRIGGER': 278,
            u'KEY_BUTTON_LEFT_THUMB_BUTTON': 276,
            u'KEY_BUTTON_LEFT_THUMB_STICK': 264,
            u'KEY_BUTTON_LEFT_THUMB_STICK_DOWN': 281,
            u'KEY_BUTTON_LEFT_THUMB_STICK_LEFT': 282,
            u'KEY_BUTTON_LEFT_THUMB_STICK_RIGHT': 283,
            u'KEY_BUTTON_LEFT_THUMB_STICK_UP': 280,
            u'KEY_BUTTON_LEFT_TRIGGER': 262,
            u'KEY_BUTTON_RIGHT_ANALOG_TRIGGER': 279,
            u'KEY_BUTTON_RIGHT_THUMB_BUTTON': 277,
            u'KEY_BUTTON_RIGHT_THUMB_STICK': 265,
            u'KEY_BUTTON_RIGHT_THUMB_STICK_DOWN': 267,
            u'KEY_BUTTON_RIGHT_THUMB_STICK_LEFT': 268,
            u'KEY_BUTTON_RIGHT_THUMB_STICK_RIGHT': 269,
            u'KEY_BUTTON_RIGHT_THUMB_STICK_UP': 266,
            u'KEY_BUTTON_RIGHT_TRIGGER': 263,
            u'KEY_BUTTON_START': 274,
            u'KEY_BUTTON_WHITE': 261,
            u'KEY_BUTTON_X': 258,
            u'KEY_BUTTON_Y': 259,
            u'KEY_INVALID': 65535,
            u'KEY_MOUSE_CLICK': 57344,
            u'KEY_MOUSE_DOUBLE_CLICK': 57360,
            u'KEY_MOUSE_DRAG': 57604,
            u'KEY_MOUSE_DRAG_END': 57606,
            u'KEY_MOUSE_DRAG_START': 57605,
            u'KEY_MOUSE_END': 61439,
            u'KEY_MOUSE_LONG_CLICK': 57376,
            u'KEY_MOUSE_MIDDLECLICK': 57346,
            u'KEY_MOUSE_MOVE': 57603,
            u'KEY_MOUSE_NOOP': 61439,
            u'KEY_MOUSE_RDRAG': 57607,
            u'KEY_MOUSE_RDRAG_END': 57609,
            u'KEY_MOUSE_RDRAG_START': 57608,
            u'KEY_MOUSE_RIGHTCLICK': 57345,
            u'KEY_MOUSE_START': 57344,
            u'KEY_MOUSE_WHEEL_DOWN': 57602,
            u'KEY_MOUSE_WHEEL_UP': 57601,
            u'KEY_UNICODE': 61952,
            u'KEY_VKEY': 61440,
            u'KEY_VMOUSE': 61439}

        '''
            u'NOTIFICATION_ERROR' : ,'error'
            u'NOTIFICATION_INFO' : ,'info'
            u'NOTIFICATION_WARNING' : ,'warning'
            u'PASSWORD_VERIFY' : 1
        '''

        remoteMap = {
            u'REMOTE_0': 58,
            u'REMOTE_1': 59,
            u'REMOTE_2': 60,
            u'REMOTE_3': 61,
            u'REMOTE_4': 62,
            u'REMOTE_5': 63,
            u'REMOTE_6': 64,
            u'REMOTE_7': 65,
            u'REMOTE_8': 66,
            u'REMOTE_9': 67,
            u'VERTICAL': 1}

        self.byNameMap = {u'actionMap': actionMap,
                          u'keyButtonMap': keyButtonMap,
                          u'remoteMap': remoteMap
                          }
        reverseActionMap = dict()
        for key in actionMap:
            value = actionMap.get(key)
            if value in reverseActionMap:
                xbmc.log(u'duplicate value in actionMap: ' +
                         str(value), xbmc.LOGDEBUG)
            reverseActionMap[value] = key

        reverseKeyButtonMap = dict()
        for key in keyButtonMap:
            value = keyButtonMap.get(key)
            if value in reverseKeyButtonMap:
                xbmc.log(u'duplicate value in keyButtonMap: ' +
                         str(value), xbmc.LOGDEBUG)
            reverseKeyButtonMap[value] = key

        reverseRemoteMap = dict()
        for key in remoteMap:
            value = remoteMap.get(key)
            if value in reverseRemoteMap:
                xbmc.log(u'duplicate value in remoteMap: ' +
                         str(value) + u' ' + reverseRemoteMap.get(value), xbmc.LOGDEBUG)
            reverseRemoteMap[value] = key

        self.reverseMapsByNameMap = {u'actionMap': reverseActionMap,
                                     u'keyButtonMap': reverseKeyButtonMap,
                                     u'remoteMap':  reverseRemoteMap}
        self.mapNames = [u'actionMap', u'keyButtonMap', u'remoteMap']

    _singletonInstance = None

    @staticmethod
    def getInstance():
        if Action._singletonInstance is None:
            Action._singletonInstance = Action()
        return Action._singletonInstance

    def getKeyIDInfo(self, action):
        actionId = action.getId()

        result = []
        for mapName in self.mapNames:
            reverseMap = self.reverseMapsByNameMap.get(mapName)
            keyName = reverseMap.get(actionId)
            if keyName is not None:
                result.append(str(actionId) + u' Map: ' +
                              mapName + u' : ' + keyName)

        if len(result) == 0:
            result.append(keyName + u' not Found')

        return result
